caproto/_hub.py

```diff
@@ -373,13 +381,27 @@ def write(self, data):
         return self.circuit, ReadNotifyRequest(data, data_type, data_count,
                                                self.sid, ioid)
 
-    def subscribe(self, data_type=None, data_count=None):
-        # TO DO
+    def subscribe(self, data_type=None, data_count=None, low=0.0, high=0.0,
+                  to=0.0, mask=None):
         if data_type is None:
             data_type = self.native_data_type
         if data_count is None:
             data_count = self.native_data_count
-        return self.circuit, SubscribeRequest(...)
+        if mask is None:
+            mask = DBE_VALUE | DBE_ALARM | DBE_PROPERTY
+
+        # TODO: hub-wide sub counter...
+        subscriptionid = next(self._hub._sub_counter)
+        self.circuit._subinfo[subscriptionid] = SubscriptionInfo(chan=self,
```

Comment:
To clarify what I was saying last night: this line mutates protocol-related state. By moving it into `_process_event`, both CLIENT and SERVER will see it on its way out/in, and thus keep an up-to-date notion of each other's state.